src/lib/gadgets/elliptic-curve.ts

```diff
@@ -0,0 +1,575 @@
+    // add in multiple of each point
+    for (let j = 0; j < n; j++) {
+      let windowSize = windowSizes[j];
+      if (i % windowSize === 0) {
+        // pick point to add based on the scalar chunk
+        let sj = scalarChunks[j][i / windowSize];
+        let sjP =
+          windowSize === 1
+            ? points[j]
+            : arrayGetGeneric(Point.provable, tables[j], sj);
+
+        // ec addition
+        let added = add(sum, sjP, Curve.modulus);
+
+        // handle degenerate case (if sj = 0, Gj is all zeros and the add result is garbage)
+        sum = Provable.if(sj.equals(0), Point.provable, sum, added);
+      }
+    }
+
+    if (i === 0) break;
+
+    // jointly double all points
+    // (note: the highest couple of bits will not create any constraints because sum is constant; no need to handle that explicitly)
+    sum = double(sum, Curve.modulus);
+  }
+
+  // the sum is now 2^(b-1)*IA + sum_i s_i*P_i
+  // we assert that sum != 2^(b-1)*IA, and add -2^(b-1)*IA to get our result
+  let iaFinal = Curve.scale(Curve.fromNonzero(ia), 1n << BigInt(b - 1));
+  Provable.equal(Point.provable, sum, Point.from(iaFinal)).assertFalse();
+  sum = add(sum, Point.from(Curve.negate(iaFinal)), Curve.modulus);
+
+  return sum;
+}
+
+/**
+ * Bigint implementation of ECDSA verify
+ */
+function verifyEcdsaConstant(
+  Curve: CurveAffine,
+  { r, s }: Ecdsa.signature,
+  msgHash: bigint,
+  publicKey: point
+) {
+  let pk = Curve.from(publicKey);
+  if (Curve.equal(pk, Curve.zero)) return false;
+  if (!Curve.isOnCurve(pk)) return false;
+  if (Curve.hasCofactor && !Curve.isInSubgroup(pk)) return false;
+  if (r < 1n || r >= Curve.order) return false;
+  if (s < 1n || s >= Curve.order) return false;
+
+  let sInv = Curve.Scalar.inverse(s);
+  assert(sInv !== undefined);
+  let u1 = Curve.Scalar.mul(msgHash, sInv);
+  let u2 = Curve.Scalar.mul(r, sInv);
+
+  let R = Curve.add(Curve.scale(Curve.one, u1), Curve.scale(pk, u2));
+  if (Curve.equal(R, Curve.zero)) return false;
+
+  return Curve.Scalar.equal(R.x, r);
+}
+
+/**
+ * Sign a message hash using ECDSA.
+ */
+function signEcdsa(Curve: CurveAffine, msgHash: bigint, privateKey: bigint) {
+  let { Scalar } = Curve;
+  let k = Scalar.random();
+  let R = Curve.scale(Curve.one, k);
+  let r = Scalar.mod(R.x);
+  let kInv = Scalar.inverse(k);
+  assert(kInv !== undefined);
+  let s = Scalar.mul(kInv, Scalar.add(msgHash, Scalar.mul(r, privateKey)));
+  return { r, s };
+}
+
+/**
+ * Given a point P, create the list of multiples [0, P, 2P, 3P, ..., (2^windowSize-1) * P].
+ * This method is provable, but won't create any constraints given a constant point.
+ */
+function getPointTable(
+  Curve: CurveAffine,
+  P: Point,
+  windowSize: number,
+  table?: Point[]
+): Point[] {
+  assertPositiveInteger(windowSize, 'invalid window size');
+  let n = 1 << windowSize; // n >= 2
+
+  assert(table === undefined || table.length === n, 'invalid table');
+  if (table !== undefined) return table;
```

Comment:
yes, so that someone could save the constraints of building the table on a variable point, by passing in the table. for constants it doesn't really matter, this is pretty quick. well for very large tables it could speed it up somewhat